modules/raft/src/main/java/org/apache/ignite/raft/jraft/rpc/impl/ActionRequestProcessor.java

```diff
@@ -82,32 +83,80 @@ public void handleRequest(RpcContext rpcCtx, ActionRequest request) {
 
         if (request instanceof WriteActionRequest) {
+            WriteActionRequest writeRequest = (WriteActionRequest)request;
+
+            WriteCommand command = writeRequest.deserializedCommand();
+
+            if (command == null) {
+                command = commandsMarshaller.unmarshall(writeRequest.command());
+            }
+
             if (fsm.getListener() instanceof BeforeApplyHandler) {
                 synchronized (groupIdSyncMonitor(request.groupId())) {
-                    callOnBeforeApply(request, fsm);
-                    applyWrite(node, (WriteActionRequest) request, rpcCtx);
+                    writeRequest = patchCommandBeforeApply(writeRequest, (BeforeApplyHandler) listener, command, commandsMarshaller);
+
+                    applyWrite(node, writeRequest, command, rpcCtx);
                 }
             } else {
-                applyWrite(node, (WriteActionRequest) request, rpcCtx);
+                applyWrite(node, writeRequest, command, rpcCtx);
             }
         } else {
-            if (fsm.getListener() instanceof BeforeApplyHandler) {
-                callOnBeforeApply(request, fsm);
+            ReadActionRequest readRequest = (ReadActionRequest) request;
+
+            if (listener instanceof BeforeApplyHandler) {
+                ReadCommand command = readRequest.command();
+
+                readRequest = patchCommandBeforeApply(readRequest, (BeforeApplyHandler) listener, command, commandsMarshaller);
             }
 
-            applyRead(node, (ReadActionRequest) request, rpcCtx);
+            applyRead(node, readRequest, rpcCtx);
         }
     }
 
-    private static void callOnBeforeApply(ActionRequest request, DelegatingStateMachine fsm) {
-        Command command = request instanceof WriteActionRequest
-                ? ((WriteActionRequest) request).command()
-                : ((ReadActionRequest) request).command();
-
-        ((BeforeApplyHandler) fsm.getListener()).onBeforeApply(command);
+    /**
+     * This method calls {@link BeforeApplyHandler#onBeforeApply(Command)} and returns action request with a serialized version of the
+     * updated command, if it has been updated. Otherwise, the method returns the original {@code request} instance. The reason for such
+     * behavior is the fact that we use {@code byte[]} in action requests, thus modified command should be serialized twice.
+     */
+    private <AR extends ActionRequest> AR patchCommandBeforeApply(
+            AR request,
+            BeforeApplyHandler beforeApplyHandler,
+            Command command,
+            Marshaller commandsMarshaller
+    ) {
+        if (beforeApplyHandler.onBeforeApply(command)) {
+            if (request instanceof WriteActionRequest) {
+                return (AR) factory.writeActionRequest()
+                    .groupId(request.groupId())
+                    .command(commandsMarshaller.marshall(command))
+                    .deserializedCommand((WriteCommand)command)
```

Comment:
Do you think you shouldn’t do this with reading commands?

Reply:
Yes, that's why I didn't do it

Reply:
Ok